packages/core/src/browser/save-resource-service.ts

```diff
@@ -58,3 +196,66 @@ export class SaveResourceService {
         return Promise.reject('Unsupported: The base SaveResourceService does not support saveAs action.');
     }
 }
+
+export class AutoSaveThrottle implements Disposable {
+
+    private _saveable: Saveable;
+    private _cb: () => void;
+    private _disposable: DisposableCollection;
+    private _throttle?: ReturnType<typeof throttle>;
+    private _mode: AutoSaveMode = 'off';
+    private _autoSaveDelay = 1000;
+
+    get autoSave(): AutoSaveMode {
+        return this._mode;
+    }
+
+    set autoSave(value: AutoSaveMode) {
+        this._mode = value;
+        this.throttledSave();
+    }
+
+    get autoSaveDelay(): number {
+        return this._autoSaveDelay;
+    }
+
+    set autoSaveDelay(value: number) {
+        this._autoSaveDelay = value;
+        // Explicitly delete the throttle to recreate it with the new delay
+        this._throttle?.cancel();
+        this._throttle = undefined;
+        this.throttledSave();
+    }
+
+    constructor(saveable: Saveable, cb: () => void, ...disposables: Disposable[]) {
+        this._cb = cb;
+        this._saveable = saveable;
+        this._disposable = new DisposableCollection(
+            ...disposables,
+            saveable.onContentChanged(() => {
+                this.throttledSave();
+            }),
+            saveable.onDirtyChanged(() => {
+                this.throttledSave();
+            })
+        );
+    }
+
+    protected throttledSave(): void {
```

Comment:
> However, autosave is for making sure I don't lose my work

You don't lose your work with `autoSave === on`, as closing the tab/window automatically saves any changes when that's enabled. This is also the convention used by VSCode, since the editor just saving between two keystrokes is bad UX.

Reply:
The question is what you want to guard against: if you want to guard against the IDE crashing, the dialog will never be shown. If you don't guard against crashes, what is autosave for?

Reply:
I don't know, why don't you ask the user that set their auto save delay to 2 minutes? ;)

IMO this way of implementing the feature is correct, especially since VSCode and Theia have been doing it the same way. I don't see a reason to deviate from that due to edge cases.

Reply:
If it's not to guard against crashes, I just don't get what "autosave" is supposed to be for. But I guess it's "works as intended".